### DISCUSSION TITLE
Use Closure on Grid::make?

### DISCUSSION THREAD

**Discussion by vv-ac** [filamentphp/filament#3096] [Help]
Hi all,

I try to have a layout that shows a Grid dependant on the status of a $resource, like this:
```
                Grid::make(function (?User $record) {
                        if (!$record->password) {
                            return ['default' => 1, 'sm' => 2];
                        }
                        else {
                            return ['default' => 1];
                        }
                    })
```

But instead of complying (:-)) it simply shows:
```
Filament\Forms\Components\Grid::make(): Argument #1 ($columns) must be of type array|int|null, Closure given, called in C:\xampp\htdocs\test\app\Filament\Resources\UserResource.php on line 60
```

As the documentation says Closures are possible on all field and layout elements, I am missing sth here, but what?

Thanks!

**Comment by ryangjchandler** [comment]
Think you want `Grid::make()->schema(function () ...)` instead.

**Reply by vv-ac** [reply]
Oh, I get it, I can just leave the definition of columns empty and control it by conditionally present different elements in the schema which then have
```
->columnSpan(function ...)?
```

This would make sense to me. I need to wait till the evening to try out (need to work unfortunately :-)).

Thank you for helping out once more!

**Reply by danharrin** [reply]
I would also accept a PR to allow the Grid to support closures for columns, but this works fine using dynamic column spans

**Comment by vv-ac** [comment]
Sorry for re-opening this one. Tonight I tried the following:
```
Grid::make(2)
                ->schema([
                    Placeholder::make('pwsection')
                        ->label('')
                        ->content(function (?User $record) {
                            if (!($record && $record->password)) {
                                return new HTMLString('...');
                            }
                            else {
                                return new HTMLString('...');
                            }
                        })
                        ->columnSpan(function (?User $record) {
                            if(!($record && $record->password)) {
                                return 'full';
                            }
                            else {
                                return 2;
                            }
                        }),

                    Toggle::make('password_change')
                        ->label('Change password')
                        ->inline(false)
                        ->onIcon('heroicon-s-shield-check')
                        ->offIcon('heroicon-s-shield-check')
                        ->reactive()
                        ->hidden(function (?User $record) {
                            return !($record && $record->password);
                        })
                        ->columnSpan(function (?User $record) {
                            if(!($record && $record->password)) {
                                return 'full';
                            }
                            else {
                                return 2;
                            }
                        }),
                    ]),
```

It does not work as intended regarding the columnSpan. The elements Placeholder and Toggle always appear below each other when the user record is already given.

Again, I am likely to oversee sth obvious?

Thanks, have a nice evening.

**Reply by danharrin** [reply]
The password field needs to be `reactive()`, and you should use `$get('password')` not `$record->password`. `$record` contains the saved hashed password, not the contents of the password field.

**Reply by vv-ac** [reply]
Thanks for your reply.
It's actually about if there already is a user created or not (simple user resource -> ManageUsers.php), so I could likely just check if $record is true.

So it's one modal that is used to create a new user but also to edit existing users.
If it's meant to be used for an existing user there is a toggle to be displayed that - if activated/set to true - should display the password field which should otherwise be hidden.
